bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -76,339 +76,296 @@
+          return;
+        }
+
+        headers.put(AUTHORIZATION_HEADER_KEY, headerCache.header);
+
+        delegate().start(new UnAuthResponseListener<>(responseListener, headerCache), headers);
       }
     };
   }
 
-  /**
-   * Refreshes the OAuth2 token asynchronously. This method will only start an async refresh if
-   * there isn't a currently running asynchronous refresh or the credentials are for App Engine.
-   * App Engine has some credentials related state in a {@link ThreadLocal} which prevents it from
-   * running asynchronously. In the App Engine case, this method will defer to
-   * {@link #syncRefresh()}.
-   *
-   * @throws IOException
-   */
-  public void asyncRefresh() throws IOException {
-    if (isAppEngine) {
-      syncRefresh();
-    } else if (canRefresh()) {
-      executor.execute(new Runnable() {
-        @Override
-        public void run() {
-          doRefresh();
-        }
-      });
+  private HeaderCacheElement getHeaderSafe() {
+    try {
+      return getHeader();
+    } catch (Exception e) {
+      return new HeaderCacheElement(
+          Status.UNAUTHENTICATED
+              .withDescription("Unexpected failure get auth token")
+              .withCause(e)
+      );
     }
   }
 
-  private boolean canRefresh() {
-    return !isRefreshing.get() && getCacheState(this.headerCache.get()) != CacheState.Good;
-  }
-
   /**
-   * <p>syncRefresh.</p>
-   *
-   * @throws java.io.IOException if any.
+   * Get the http credential header we need from a new oauth2 AccessToken.
    */
-  public void syncRefresh() throws IOException {
-    synchronized (isRefreshing) {
-      if (!isRefreshing.get()) {
-        doRefresh();
+  private HeaderCacheElement getHeader() throws ExecutionException, InterruptedException {
+    final Future<HeaderCacheElement> deferredResult;
+
+    synchronized (lock) {
+      CacheState state = headerCache.getCacheState();
+
+      if (state == CacheState.Good || state == CacheState.Exception) {
+        return headerCache;
+        // AppEngine doesn't allow threads to outlive the request, so disable background refresh
+      } else if (!isAppEngine && state == CacheState.Stale) {
+        asyncRefresh();
+        return headerCache;
+      } else if (state == CacheState.Expired) {
+        // defer the future resolution (asyncRefresh will spin up a thread that will try to acquire the lock)
+        deferredResult = asyncRefresh();
       } else {
-        while (isRefreshing.get() && getCacheState(this.headerCache.get()) != CacheState.Good) {
-          try {
-            isRefreshing.wait(250);
-          } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-            throw new IOException(e);
-          }
-        }
+        return new HeaderCacheElement(
+            Status.UNAUTHENTICATED
+                .withCause(new IllegalStateException("Could not process state: " + state))
+        );
       }
     }
+    return deferredResult.get();
```

Comment:
This should timeout after some amount of time, probably 250 ms. like the sync refresh.